core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2345,17 +2345,19 @@ private[spark] class RedirectThread(
  */
 private[spark] class CircularBuffer(sizeInBytes: Int = 10240) extends java.io.OutputStream {
   var pos: Int = 0
+  var isBufferFull = false
   var buffer = new Array[Int](sizeInBytes)
 
   def write(i: Int): Unit = {
     buffer(pos) = i
     pos = (pos + 1) % buffer.length
+    isBufferFull = isBufferFull || (pos == 0)
   }
 
   override def toString: String = {
     val (end, start) = buffer.splitAt(pos)
     val input = new java.io.InputStream {
-      val iterator = (start ++ end).iterator
+      val iterator = (if (isBufferFull) start ++ end else end).iterator
```

Comment:
Again I know it's not your code but I also don't get why this code reads the stream as a stream of chars just to make it into a string. It can be made into a string directly. The only possible difference this makes is to convert the various line-ending chars that `BufferedReader` understands into `\n`, but that can be done more efficiently and explicitly -- if that was the intent at all!